syntax = "proto3";

option go_package = "./raft";

package raft;

message Log {
    string o = 1;   // current log operation
    int32 t = 2;    // term associated with this log
    int32 k = 3;    // index of this log in the leader's logs
}

message AppendEntriesRequest {
    string leaderId = 1;    // so that followers know who is the current leader
    int32 c = 2;            // The index of the most recent commited operation
    repeated Log logs = 3;  // simplified version: leader send the entires logs
}

message AppendEntriesResponse {
    bool success = 1;
}

message RequestVoteRequest {
    int32 term = 1;         // candidate's term
    string candidateId = 2; // id of candidate requesting RequestVote
}

message RequestVoteResponse {
    bool voteGranted = 1;
}

message ReportRequest {
    string timeStamp = 1;
    string rpcType = 2;
    string sender = 3;
    string receiver = 4;
    string action = 5;
}

message ReportResponse {
    bool success = 1;
}

service Raft {
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
}

service Report {
    rpc SendReport(ReportRequest) returns (ReportResponse);
}

